### DISCUSSION TITLE
How do I use breakpoints in Godot-jolt?

### DISCUSSION THREAD

**Discussion by Ewall198** [godot-jolt/godot-jolt#975] [Questions]
I'm using VSCode & Godot 4.3-stable
I have the Godot engine compiling from source and set up to run my scene (no project select or editor). I know that breakpoints are working in godot, because I can set a breakpoint in the source code of Godot engine.

Then I do a fresh build of Godot-Jolt with debugging symbols. But, I'm not able to hit the breakpoints that I put in Godot-Jolt, I put a bunch of different breakpoints in that are definitely getting hit, like the first line of `register_types.cpp:on_initialize`. Any idea what I'm doing wrong?

Below are my config files.

`godot\.vscode\tasks.json`
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "build",
      "group": "build",
      "type": "shell",
      "command": "scons",
      "args": [
        "dev_build=yes",
        "target=template_debug",
        "dev_mode=yes",
        "debug_symbols=yes",
        "optimize=none"
      ],
      "problemMatcher": "$msCompile"
    }
  ]
}
```

***
`godot\.vscode\launch.json`
```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Launch Project",
      "type": "cppvsdbg",
      "request": "launch",
      "program": "${workspaceFolder}/bin/godot.windows.editor.dev.x86_64.exe",
      "args": [
        "--debug",
        "--clean",
        "--path",
        "E:/Users/ewall/Godot/engine/godot-jolt-forked/examples",
        "res://scenes/joints/entities/distance/distance.tscn"
        // "--editor"
      ],
      "stopAtEntry": false,
      "cwd": "${workspaceFolder}",
      "environment": [],
      "console": "internalConsole",
      "visualizerFile": "${workspaceFolder}/platform/windows/godot.natvis",
      "preLaunchTask": "build"
    }
  ]
}
```

***
`godot-jolt-forked\CMakeUserPresets.json`
```json
{
  "version": 3,
  "configurePresets": [
    {
      "name": "custom-dev",
      "inherits": [
        "windows-msvc-x64",
        "dev-base"
      ],
      "displayName": "MSVC, 64-bit, Development",
      "cacheVariables": {
        "GDJ_INSTALL_DEBUG_SYMBOLS": {
          "type": "BOOL",
          "value": true
        },
        "GDJ_INTERPROCEDURAL_OPTIMIZATION": {
          "type": "BOOL",
          "value": false
        }
      }
    }
  ],
  "buildPresets": [
    {
      "name": "custom-debug-editor",
      "configurePreset": "custom-dev",
      "configuration": "EditorDebug",
      "displayName": "CustomEditorDebug",
      "targets": [
        "install"
      ]
    },
    {
      "name": "custom-debug",
      "configurePreset": "custom-dev",
      "configuration": "Debug",
      "displayName": "CustomDebug",
      "targets": [
        "install"
      ]
    }
  ]
}
```

**Comment by mihe** [comment]
Some minor notes that may or may not be related:

1. You don't need to use `GDJ_INSTALL_DEBUG_SYMBOLS`. When you build a non-distribution configuration (like `EditorDebug`) the PDB path in the DLL header will point to the absolute path of the PDB at `[repo]/build/[preset]/[config]/[library].pdb`, so the PDB doesn't need to sit next to the DLL like with the official distributions.
2. You don't need to pass the project path as `--path`, you can just set it as the `cwd` instead.
3. You shouldn't need to pass `visualizerFile` as it gets baked into the Godot PDB during linking.
4. Your user presets are all non-optimized, so `GDJ_INTERPROCEDURAL_OPTIMIZATION` won't be used and is redundant.
5. `optimize=none` is redundant for `dev_build=yes` when building Godot.

> I put a bunch of different breakpoints in that are definitely getting hit, like the first line of `register_types.cpp:on_initialize`

I've found the `cppvsdbg` debugger in VS Code to be a bit strange when it comes to breakpoints. It doesn't seem to care at all about file paths, but only their file names, so if you place a breakpoint anywhere in `register_types.cpp` you will effectively break on that line in _all_ files by that name, and there are quite a few `register_types.cpp` in Godot, so I would probably try putting a breakpoint in something like `JoltPhysicsServer3D` constructor instead, which should run pretty early in the flow as well.

If you place a breakpoint in `JoltPhysicsServer3D`, does it actually turn red, or does it turn gray, and if it does turn gray, what does it say if you hover over it? Typically it'll say something like "No symbols loaded for this document" if there's something wrong with the symbol loading.

If it doesn't turn gray, but instead red, and just never hits the breakpoint, then I'm at a loss for advice, unfortunately. At that point your best bet is to put a breakpoint in `PhysicsServer3DManager::new_default_server` on the Godot side of things and step from there and see where you end up.

Besides all that I don't see anything that sticks out in your setup. It looks fairly similar to my own.
